### PR TITLE
pulsar support

### PR DESCRIPTION
**Pull request by venkatesh-sharma17**

### CODE REVIEW

File: core/dmux.go [hunk 1]
```diff
@@ -220,7 +220,7 @@ func getStopMsg() ControlMsg {
 	return c
 }
 
-func (d *Dmux) runWithSideline(source Source, sink Sink, sidelineImpl sideline_module.CheckMessageSideline) {
+func (d *Dmux) runWithSideline(source Source, sink Sink, sidelineImpl sideline_module.CheckMessageSideline, enableDebugLog bool) {
```
Comment (kingster) [inline core/dmux.go]: Instead of adding another parameter, can we make it a struct of optional params?

File: pulsar/message.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+)
+
+// Message is a container of message and it's state
+type Message struct {
+	Msg       *pulsar.ConsumerMessage
+	Processed bool
+	Sidelined bool
+}
+
+// GetPayload implements HTTPMsg interface
```
Comment (kingster) [inline pulsar/message.go]: Does this implement HTTPMsg or this is a typo?

File: pulsar/message.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+		strconv.FormatInt(m.Msg.ID().EntryID(), 10),
+		strconv.FormatInt(int64(m.Msg.ID().BatchIdx()), 10),
+		strconv.FormatInt(int64(m.Msg.ID().PartitionIdx()), 10))
+}
+
+// GetURL implements HTTPMsg interface
+func (m *Message) GetURL(endpoint string) string {
+	var builder strings.Builder
+	_topic := strings.Split(m.Msg.Topic(), "/")
+	topic := _topic[len(_topic)-1]
```
Comment (kingster) [inline pulsar/message.go:38]: Why do we need to do something like this? Can we not create using any better way?

File: core/dmux.go (hunk 1 above)
Comment (venkatesh-sharma17) [reply core/dmux.go]: have added optional params, please check now

File: core/dmux.go [hunk 4]
```diff
@@ -48,26 +48,30 @@ type Sideline struct {
 	SidelineMeta          interface{} `json:"sidelineMeta"`
 }
 
+type OptionalParamsForDMUXStart struct {
```
Comment (kingster) [inline core/dmux.go]: Just name this as DmuxOptionalParams

File: go.mod [hunk 5]
```diff
@@ -1,30 +1,21 @@
 module github.com/flipkart-incubator/go-dmux
 
-go 1.12.4
+go 1.12
```
Comment (kingster) [inline go.mod]: probably bump the go version. 1.12 is very old.

File: http/http_sink.go [hunk 6]
```diff
@@ -169,6 +169,8 @@ func (h *HTTPSink) Consume(msg interface{}, retries int, sidelineResponseCodes [
 func (h *HTTPSink) retryPre(msg interface{}, url string) {
 	for {
 		status := h.pre(h.hook, msg, url)
+		log.Print("url is " + url + "  status is " + strconv.FormatBool(status))
```
Comment (kingster) [inline http/http_sink.go]: this will log a lot, is this expected?

File: pulsar/source.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+func (p *PulsarSource) GetValue(msg interface{}) []byte {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (p *PulsarSource) GetOffset(msg interface{}) int64 {
+	//TODO implement me
+	panic("implement me")
+}
```
Comment (kingster) [inline pulsar/source.go:38]: any reason to keep this unimplemented?

File: pulsar/source.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+func (p *PulsarSource) Generate(out chan<- interface{}) {
+	// Prepare KeyFile
+	props := map[string]string{
+		"type":          "client_credentials",
+		"client_id":     p.conf.AuthClientId,
+		"client_secret": p.conf.AuthClientSecret,
+		"issuer_url":    p.conf.AuthIssuerURL,
+	}
+	privateKey, _ := json.Marshal(props)
+	fmt.Println("prepared pulsar privateKey", privateKey)
```
Comment (kingster) [inline pulsar/source.go]: remove this log. This is a secret which shouldn't be logged

File: pulsar/source.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+	// Build authentication
+	auth := pulsar.NewAuthenticationOAuth2(map[string]string{
+		"type":       "client_credentials",
+		"issuerUrl":  p.conf.AuthIssuerURL,
+		"audience":   p.conf.AuthAudience,
+		"privateKey": fmt.Sprintf("data://%s", string(privateKey)),
+		"clientId":   p.conf.AuthClientId,
+	})
+	fmt.Println("prepared pulsar oauth2")
```
Comment (kingster) [inline pulsar/source.go]: remove these fmt.Println debug print to stdout

File: pulsar/source.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+		log.Println("Done setting force restart as true and readnewest as true")
+	} else if p.conf.ForceRestart && p.conf.SeekByTime != 0 {
+		seekByTime := time.Duration(p.conf.SeekByTime) * time.Second
+		updatedTime := time.Now().UTC().Add(-seekByTime)
+		er := consumer.SeekByTime(updatedTime)
+		if er != nil {
+			client.Close()
+			panic(er)
+		}
+		fmt.Printf("Done Setting SeekByTime as %v %v\n", seekByTime.String(), updatedTime)
```
Comment (kingster) [inline pulsar/source.go]: Change to log.Println. Don't mix fmt and log.

File: pulsar/source.go (hunk 7 above)
Comment (venkatesh-sharma17) [reply pulsar/source.go:38]: all these methods are getting used in sideline related flow, and need to be carefully implemented..And as we have have decided to limit the scope of Pulsar to non-sideline enabled dmux, these methods are not implemented yet.. shall we implement these?

File: pulsar/source.go (hunk 8 above)
Comment (venkatesh-sharma17) [reply pulsar/source.go]: ack

File: pulsar/source.go (hunk 9 above)
Comment (venkatesh-sharma17) [reply pulsar/source.go]: ack

File: pulsar/source.go (hunk 10 above)
Comment (venkatesh-sharma17) [reply pulsar/source.go]: ack

File: http/http_sink.go (hunk 6 above)
Comment (venkatesh-sharma17) [reply http/http_sink.go]: ack, will remove

File: go.mod (hunk 5 above)
Comment (venkatesh-sharma17) [reply go.mod]: ack, changing it to 1.13.. due to some reason the 1.12.4 was giving error
" invalid go version '1.12.4': must match format 1.23"

File: core/dmux.go (hunk 4 above)
Comment (venkatesh-sharma17) [reply core/dmux.go]: ack

File: pulsar/message.go (hunk 2 above)
Comment (venkatesh-sharma17) [reply pulsar/message.go]: ack, yes it was a typo, Message here implements HTTPmessage

File: pulsar/message.go (hunk 3 above)
Comment (venkatesh-sharma17) [reply pulsar/message.go:38]: pulsar topics are fully qualified names like "persistent://abc/def/TOPIC_NAME.. to send the entire FQN would be too much info in the URI.. we needed to extract the last identifier that  i.e., topic name, hence had to use split and all

File: connection/pulsar_http_conn.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+type PulsarConn struct {
+	EnableDebugLog bool
+	Conf           interface{}
+}
+
+// getConfiguration parses configs and returns connection config
+func (c *PulsarConn) getConfiguration() *PulsarConnConfig {
+	data, _ := json.Marshal(c.Conf)
+	var config *PulsarConnConfig
+	json.Unmarshal(data, &config)
```
Comment (kingster) [inline connection/pulsar_http_conn.go:29]: Why marshal and unmarshal again? Can please change the type of Conf to PulsarConnConfig to remove this

File: java/godmux-tools/src/test/java/com/flipkart/godmux/tools/BatchDecoderTest.java [hunk 12]
```diff
@@ -26,7 +26,7 @@ public void testBatchDecoding() throws IOException {
         byte[] payload = Arrays.copyOfRange(data, 0, counter);
         byte[][] output = BatchDecoder.DECODE.decode(payload);
         for (int i = 0; i < output.length; i++) {
-            System.out.println(new String(output[i]));
+            System.out.Println(new String(output[i]));
```
Comment (kingster) [inline java/godmux-tools/src/test/java/com/flipkart/godmux/tools/BatchDecoderTest.java:29]: typo??

File: pulsar/hook.go [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+	if h.enableDebugLog {
+		data := msg.(sink.HTTPMsg)
+		log.Printf("%s before http sink \n", data.GetDebugPath())
+	}
+}
+
+// PostHTTPCall is invoked - after HttpSink execution. This implementation calls
+// KafkaMessage MarkDone method on the data argument of Post, to mark this
+// message and successfully processed.
+func (h *CursorHook) PostHTTPCall(msg interface{}, success bool) {
```
Comment (kingster) [inline pulsar/hook.go:52]: Can this me discret type instead of `interface{}`

File: pulsar/message.go [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+	header["Content-Type"] = "application/json" //force json for foxtrot
+
+	return header
+}
+
+// CustomURLKey  place holder name, which will be replaced by kafka key
+const CustomURLKey = "__KEY_NAME__"
+
+// BatchURL implements HTTPMsg interface
+func (m *Message) BatchURL(msgs []interface{}, endpoint string, version int) string {
```
Comment (kingster) [inline pulsar/message.go:64]: Can this me discret type instead of `interface{}`

**Review by kingster — COMMENTED** [review]
Mostly looks okay. Is it possible to change `interface{}` to discrete types in all places to avoid typecasting ?

**Review by tesrohit-developer — APPROVED** [review]
LGTM

**Pull request merged**